jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,486 @@
+     * <p>{@link Duration#ZERO} means that the request may stay suspended forever.</p>
+     *
+     * @param maxSuspend the max duration of time a request may stay suspended
+     */
+    public void setMaxSuspend(Duration maxSuspend)
+    {
+        if (maxSuspend.isNegative())
+            throw new IllegalArgumentException("invalid maxSuspend duration");
+        this.maxSuspend = maxSuspend;
+    }
+
+    @ManagedAttribute("The number of suspended requests")
+    public long getSuspendedRequestCount()
+    {
+        return state.getLo();
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        timeouts = new Timeouts(getServer().getScheduler());
+        addBean(timeouts);
+
+        int maxRequests = getMaxRequestCount();
+        if (maxRequests <= 0)
+        {
+            ThreadPool threadPool = getServer().getThreadPool();
+            if (threadPool instanceof ThreadPool.SizedThreadPool sized)
+                maxRequests = sized.getMaxThreads() / 2;
+            else
+                maxRequests = ProcessorUtils.availableProcessors();
+            setMaxRequestCount(maxRequests);
+        }
+        state.set(maxRequests, 0);
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} initialized maxRequests={}", this, maxRequests);
+
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        removeBean(timeouts);
+        timeouts.destroy();
+    }
+
+    @Override
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} handling {}", this, request);
+
+        while (true)
+        {
+            long encoded = state.get();
+            int permits = AtomicBiInteger.getHi(encoded);
+            if (permits > 0)
+            {
+                if (!state.compareAndSet(encoded, permits - 1, 0))
+                    continue;
+                return forward(request, response, callback);
+            }
+            else
+            {
+                if (request.getAttribute(EXPIRED_ATTRIBUTE_NAME) != null)
+                {
+                    notAvailable(response, callback);
+                }
+                else
+                {
+                    // Avoid this race condition:
+                    // T1 in handle() may find no permits, so it will suspend the request.
+                    // T2 in resume() finds no suspended requests and increments the permits.
+                    // T1 suspends the request, which will remain suspended despite permits are available.
+                    // Below, the suspended request count is only incremented if there are no permits.
+                    // See correspondent code in resume() and expire().
+                    int suspended = AtomicBiInteger.getLo(encoded);
+                    if (!state.compareAndSet(encoded, 0, suspended + 1))
```

Comment:
This is OK with the current fragile code.  But with either of the alternatives I suggest above, this needs to be:
```suggestion
                    if (!state.compareAndSet(encoded, permits, suspended + 1))
```

I know you like the `0`s, but I actually prefer seeing `state.compareAndSet(encoded, function(permits), function(suspended)` as hard coded values like `0` are just assuming the value in encoded and are thus fragile.

Reply:
It's not fragile, is it explicitating the invariant.

Reply:
@sbordet It is not making invariant explicit.  Instead it is assuming the invariant and hence it is fragile.  If the invariant ever changes, then this code will break.  To make the invariant explicit it needs to extract the permits field and check that it is zero.   Currently is just overwriting whatever the value is to zero because it assumes it is zero.

Maybe a single AtomicInteger would be better, with positive values representing permits and negative values representing the suspended count.   Then the invariant could never be broken as you could not have both permits and suspended at the same time.   If your state variables allow illegal states, then you need to check that those states are not entered rather than assume they are not.